src/backupNg/backupNg.spec.js

```diff
@@ -234,4 +224,119 @@ describe("backupNg", () => {
       expect(vmTask.data.id).toBe(vmIdWithoutDisks);
     });
   });
+
+  test("execute three times a rolling snapshot with 2 as retention & revert to an old state", async () => {
+    jest.setTimeout(7e4);
+    const vmId = await xo.createTempVm({
+      name_label: "XO Test Temp",
+      name_description: "Creating a temporary vm",
+      template: config.xoTestTemplateId,
+      VDIs: [
+        {
+          size: 1,
+          SR: config.srs.defaultSr,
+          type: "user",
+        },
+      ],
+    });
+
+    const scheduleTempId = randomId();
+    const { id: jobId } = await xo.createTempBackupNgJob({
+      ...defaultBackupNg,
+      vms: {
+        id: vmId,
+      },
+      schedules: {
+        [scheduleTempId]: DEFAULT_SCHEDULE,
+      },
+      settings: {
+        ...defaultBackupNg.settings,
+        [scheduleTempId]: { snapshotRetention: 2 },
+      },
+    });
+
+    const schedule = await xo.getSchedule({ jobId });
+    expect(typeof schedule).toBe("object");
+    for (let i = 0; i < 3; i++) {
+      const oldSnapshots = xo.objects.all[vmId].snapshots;
+      await xo.call("backupNg.runJob", { id: jobId, schedule: schedule.id });
+      await xo.waitObjectState(vmId, ({ snapshots }) => {
+        // Test on updating snapshots.
+        expect(snapshots).not.toEqual(oldSnapshots);
+      });
+    }
+
+    const { snapshots, videoram: oldVideoram } = xo.objects.all[vmId];
+
+    // Test on the retention, how many snapshots should be saved.
+    expect(snapshots.length).toBe(2);
+
+    const newVideoram = 16;
+    await xo.call("vm.set", { id: vmId, videoram: newVideoram });
+    await xo.waitObjectState(vmId, ({ videoram }) => {
+      expect(videoram).toBe(newVideoram.toString());
+    });
+
+    await xo.call("vm.revert", {
+      snapshot: snapshots[0],
+    });
+
+    await xo.waitObjectState(vmId, ({ videoram }) => {
+      expect(videoram).toBe(oldVideoram);
+    });
+
+    const [
+      {
+        tasks: [
+          {
+            tasks: [task1, task2, task3],
+            ...vmTask
+          },
+        ],
+        ...log
+      },
+    ] = await xo.call("backupNg.getLogs", {
+      jobId,
+      scheduleId: schedule.id,
+    });
+
+    expect(log).toMatchSnapshot({
+      end: expect.any(Number),
+      id: expect.any(String),
+      jobId: expect.any(String),
+      scheduleId: expect.any(String),
+      start: expect.any(Number),
+    });
+
+    expect(vmTask).toMatchSnapshot({
+      data: {
+        id: expect.any(String),
+      },
+      end: expect.any(Number),
+      id: expect.any(String),
+      message: expect.any(String),
+      start: expect.any(Number),
+    });
+
+    expect(task1).toMatchSnapshot({
```

Comment:
@julien-f I want your opinion about the log validation

We want to validate the generated logs on the backup execution. My question is, should we also validate the tasks which are not handled by XO ?

Reply:
IMHO you should only validate documented tasks, those are part of the API, the others are here to help debugging but are not part of the contract.

Reply:
@julien-f  Thank you for your response :+1: